File: android-vinetrack/app/src/main/java/com/rork/vinetrack/data/material/WorkTaskMaterialCostsAccess.kt
package com.rork.vinetrack.data.material

import com.rork.vinetrack.data.auth.SessionPhase

/**
 * THE single access decision for Work Task Material Costs (sql/247).
 *
 * Work Task materials follow the existing vineyard Work Task access path.
 *
 * Material Library entry points retain their separate System Admin restriction.
 *
 * ## Why the database is NOT gated on System Admin
 *
 * sql/247 deliberately uses the ordinary vineyard/work-task policies:
 * `is_vineyard_member` for reads, `has_vineyard_role` for writes. Baking
 * "System Admin only" into RLS would introduce a platform-authority
 * carve-out inside a tenant's costing data.
 *
 * ## This is presentation only
 *
 * A cached client boolean is not authorisation. Every read and write is
 * independently enforced by RLS; hiding an entry point protects nobody on its
 * own. Mirrors the iOS `WorkTaskMaterialCostsAccess.swift` exactly.
 */
sealed interface WorkTaskMaterialCostsAccess {

    /** Material Costs may be shown and exercised. */
    data object Allowed : WorkTaskMaterialCostsAccess

    /** Everyone else, including unauthenticated and still-resolving sessions. */
    data class Unavailable(val reason: Reason) : WorkTaskMaterialCostsAccess

    /**
     * Why access was refused. Diagnostic only.
     */
    enum class Reason {
        /** No authenticated session (signed out). */
        NotAuthenticated,

        /** Session still hydrating; nothing about the user is known yet. */
        SessionRestoring,


        /**
         * No vineyard is selected, or the caller is not a member of it.
         * Material Costs is vineyard data; platform authority never crosses a
         * tenancy boundary. This reason SURVIVES removal of the gate.
         */
        NotVineyardMember,
    }

    val isAllowed: Boolean get() = this is Allowed

    companion object {
        /**
         * Resolve the single decision.
         *
         * Fail-closed: every unresolved state lands in [Unavailable], so a
         * Material Costs entry point can never flash into view during launch
         * for someone who does not have it.
         */
        fun resolve(
            sessionPhase: SessionPhase,
            selectedVineyardId: String?,
            isMemberOfSelectedVineyard: Boolean,
        ): WorkTaskMaterialCostsAccess = when {
            sessionPhase == SessionPhase.Restoring -> Unavailable(Reason.SessionRestoring)
            !sessionPhase.isAuthenticated -> Unavailable(Reason.NotAuthenticated)
            selectedVineyardId.isNullOrBlank() -> Unavailable(Reason.NotVineyardMember)
            !isMemberOfSelectedVineyard -> Unavailable(Reason.NotVineyardMember)
            else -> Allowed
        }

    }
}

File: ios/VineTrack/App/WorkTaskMaterialCostsAccess.swift
import Foundation

/// THE single access decision for Work Task Material Costs (sql/247).
///
/// Material lines use the existing Work Task access path; the Material Library
/// in Settings retains its separate System Admin restriction.
///
/// ## Why the database is NOT gated on System Admin
///
/// sql/247 deliberately uses the ordinary vineyard/work-task policies:
/// `is_vineyard_member` for reads, `has_vineyard_role` for writes. Baking
/// "System Admin only" into RLS would introduce a platform-authority
/// carve-out in a tenant's costing data.
///
/// ## This is presentation only
///
/// A cached client boolean is not authorisation. Every read and write is
/// independently enforced by RLS; hiding an entry point protects nobody on its
/// own. Mirrors the Android `WorkTaskMaterialCostsAccess.kt`.
nonisolated enum WorkTaskMaterialCostsAccess: Equatable, Sendable {

    /// Material Costs may be shown and exercised.
    case allowed

    /// Everyone else, including unauthenticated and still-resolving sessions.
    case unavailable(Reason)

    /// Why access was refused. Diagnostic only.
    nonisolated enum Reason: Equatable, Sendable {
        /// No authenticated session (signed out).
        case notAuthenticated
        /// Session still resolving; nothing is known yet.
        case stillResolving
        /// No vineyard is selected, or the caller is not a member of it.
        /// Material Costs is vineyard data; platform authority never crosses a
        /// tenancy boundary. This reason SURVIVES removal of the gate.
        case notVineyardMember
    }

    var isAllowed: Bool { self == .allowed }

    /// Resolve the single decision.
    ///
    /// Fail-closed: every unresolved state lands in `unavailable`, so a
    /// Material Costs entry point can never flash into view during launch for
    /// someone who does not have it.
    ///
    /// - Parameters:
    ///   - isAuthenticated: a signed-in session exists.
    ///   - isResolving: the session has not settled.
    ///   - selectedVineyardID: the vineyard currently in context.
    ///   - isMemberOfSelectedVineyard: the caller holds a role in it.
    static func resolve(
        isAuthenticated: Bool,
        isResolving: Bool,
        selectedVineyardID: UUID?,
        isMemberOfSelectedVineyard: Bool
    ) -> WorkTaskMaterialCostsAccess {
        if isResolving { return .unavailable(.stillResolving) }
        if !isAuthenticated { return .unavailable(.notAuthenticated) }
        guard selectedVineyardID != nil else { return .unavailable(.notVineyardMember) }
        if !isMemberOfSelectedVineyard { return .unavailable(.notVineyardMember) }
        return .allowed
    }

}
